com_msg.c

```diff
@@ -39,13 +39,13 @@ float BuffBigFloat (const unsigned char *buffer)
 		unsigned int i;
 	}
 	u;
-	u.i = (buffer[0] << 24) | (buffer[1] << 16) | (buffer[2] << 8) | buffer[3];
+	u.i = ((unsigned)buffer[0] << 24) | (buffer[1] << 16) | (buffer[2] << 8) | buffer[3];
 	return u.f;
 }
 
 int BuffBigLong (const unsigned char *buffer)
 {
-	return (buffer[0] << 24) | (buffer[1] << 16) | (buffer[2] << 8) | buffer[3];
+	return ((unsigned)buffer[0] << 24) | (buffer[1] << 16) | (buffer[2] << 8) | buffer[3];
```

Comment:
Could you include comment(s) please about why these specific chars need to be cast? Is there a platform-specific issue?

Reply:
The integer promotions apply to the left-hand side of the left-shift operator (C99, 6.5.7.3), so `buffer[0]` is converted to `int` (C99, 6.3.1.1.2).  DP seems to be written assuming `int` is 32 bits. If the result of the shift would be unrepresentable in `int`, which happens if `buffer[0]` is greater 127, UB is invoked (C99, 6.5.7.4). I'll add an explanatory comment.